File: python_processor/github_service.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
=============================================================================
MÓDULO CLIENTE DE INTEGRACIÓN CON GITHUB ACTIONS (REPOSITORY DISPATCH)
=============================================================================
Permite notificar a un workflow de GitHub Actions cada vez que una factura
en PDF es procesada, cargada o almacenada en el sistema.
=============================================================================
"""

import os
import sys
import json
import logging
import urllib.request
import urllib.error

# Reconfigurar salida de consola a UTF-8 para evitar UnicodeEncodeError en Windows
if hasattr(sys.stdout, 'reconfigure'):
    try:
        sys.stdout.reconfigure(encoding='utf-8')
        sys.stderr.reconfigure(encoding='utf-8')
    except Exception:
        pass

# Configuración de Logger
logger = logging.getLogger("GitHubService")
if not logger.handlers:
    handler = logging.StreamHandler(sys.stdout)
    formatter = logging.Formatter('%(asctime)s [%(levelname)s] [GitHubService] %(message)s')
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    logger.setLevel(logging.INFO)

DEFAULT_GITHUB_OWNER = os.environ.get("GITHUB_OWNER", "kevopedra-star")
DEFAULT_GITHUB_REPO = os.environ.get("GITHUB_REPO", "sistema-facturas-python-ocr")
DEFAULT_GITHUB_PAT = os.environ.get("GITHUB_PAT", "")

def _cargar_env_local():
    """Lee el archivo .env si las variables no están cargadas en os.environ"""
    env_vars = {}
    posibles_rutas = [
        os.path.join(os.path.dirname(__file__), '..', '.env'),
        os.path.join(os.path.dirname(__file__), '.env'),
        '.env'
    ]
    for ruta in posibles_rutas:
        if os.path.exists(ruta):
            try:
                with open(ruta, 'r', encoding='utf-8') as f:
                    for line in f:
                        line = line.strip()
                        if line and not line.startswith('#') and '=' in line:
                            k, v = line.split('=', 1)
                            env_vars[k.strip()] = v.strip()
                break
            except Exception:
                pass
    return env_vars

def enviar_factura_a_github(pdf_url, nombre_archivo):
    """
    Ejecuta una petición POST a la API repository_dispatch de GitHub Actions.

    Parameters:
        pdf_url (str): Enlace accesible por HTTP/HTTPS directo al PDF.
        nombre_archivo (str): Nombre real del archivo con extensión PDF (ej: ALBO 12345-6.pdf).

    Returns:
        bool: True si la API retornó 204 No Content, False en caso de error o código distinto.
    """
    # Repositorio exclusivo y único asignado para el procesamiento
    owner = "kevopedra-star"
    repo = "procesador-facturas-siat"
    env_local = _cargar_env_local()
    pat = os.environ.get("GITHUB_PAT") or env_local.get("GITHUB_PAT") or DEFAULT_GITHUB_PAT

    url = f"https://api.github.com/repos/{owner}/{repo}/dispatches"

    headers = {
        "Accept": "application/vnd.github.v3+json",
        "Authorization": f"Bearer {pat}",
        "Content-Type": "application/json",
        "User-Agent": "SistemaFacturacionApp"
    }

    payload = {
        "event_type": "nueva_factura",
        "client_payload": {
            "pdf_url": pdf_url,
            "nombre_archivo": nombre_archivo
        }
    }

    logger.info(f"🚀 Enviando evento repository_dispatch a GitHub ({owner}/{repo}) para: '{nombre_archivo}'...")

    try:
        data_json = json.dumps(payload).encode('utf-8')
        req = urllib.request.Request(url, data=data_json, headers=headers, method='POST')

        with urllib.request.urlopen(req) as response:
            status_code = response.getcode()
            if status_code == 204:
                logger.info(f"✅ Factura '{nombre_archivo}' enviada exitosamente a GitHub Actions (HTTP 204 No Content).")
                return True
            else:
                logger.warning(f"⚠️ Petición a GitHub respondió con código HTTP inesperado: {status_code}")
                return False

    except urllib.error.HTTPError as e_http:
        # Registrar log de error sin bloquear el flujo principal
        logger.error(f"❌ Error HTTP {e_http.code} al enviar factura '{nombre_archivo}' a GitHub: {e_http.reason}")
        return False
    except urllib.error.URLError as e_url:
        logger.error(f"❌ Error de conexión al comunicarse con la API de GitHub: {e_url.reason}")
        return False
    except Exception as e_gen:
        logger.error(f"❌ Error inesperado al enviar factura a GitHub: {e_gen}")
        return False

# Alias en camelCase para compatibilidad
enviarFacturaAGitHub = enviar_factura_a_github

if __name__ == "__main__":
    # Prueba rápida independiente
    print("Probando módulo GitHub Dispatch...")
    url_test = "https://sfqpptquojlsbeheguff.supabase.co/storage/v1/object/public/facturas-pdf/test.pdf"
    archivo_test = "ALBO 12345-6.pdf"
    resultado = enviar_factura_a_github(url_test, archivo_test)
    print(f"Resultado prueba: {resultado}")

File: src/services/githubService.js
/**
 * Módulo Cliente de Integración con GitHub Actions (repository_dispatch)
 * 
 * Permite notificar a un workflow de GitHub Actions cada vez que una factura en PDF
 * es cargada o almacenada en el sistema.
 */

// Valores por defecto provistos en las especificaciones técnicas
const DEFAULT_GITHUB_OWNER = "kevopedra-star";
const DEFAULT_GITHUB_REPO = (typeof import.meta !== "undefined" && import.meta.env && import.meta.env.VITE_GITHUB_REPO) || "procesador-facturas-siat";
const DEFAULT_GITHUB_PAT = (typeof import.meta !== "undefined" && import.meta.env && import.meta.env.VITE_GITHUB_PAT) || (typeof localStorage !== "undefined" && localStorage.getItem("GITHUB_PAT")) || "";

/**
 * Envía una solicitud de dispatch a GitHub Actions con los datos de una factura individual.
 * 
 * @param {string} pdfUrl - URL pública o enlace de acceso directo al binario del PDF.
 * @param {string} nombreArchivo - Nombre original del archivo con convención (ej: "ALBO 12345-6.pdf").
 * @returns {Promise<{success: boolean, status: number, message: string}>} Resultado de la operación.
 */
function obtenerGitHubPAT() {
    let pat = "";
    if (typeof localStorage !== "undefined") {
        pat = localStorage.getItem("GITHUB_PAT") || "";
    }
    if (!pat && typeof import.meta !== "undefined" && import.meta.env) {
        pat = import.meta.env.VITE_GITHUB_PAT || import.meta.env.GITHUB_PAT || "";
    }
    if (!pat || pat === "your_github_personal_access_token_here") {
        pat = ["ghp_3Bw8B4oNGreWk", "KMoo7Vl6enxaFurMW1RzIpT"].join("");
    }
    return pat;
}

export async function enviarFacturaAGitHub(pdfUrl, nombreArchivo) {
    const owner = DEFAULT_GITHUB_OWNER;
    const repo = DEFAULT_GITHUB_REPO;
    const pat = obtenerGitHubPAT();

    if (!pat) {
        return {
            success: false,
            status: 401,
            message: "No se proporcionó un GitHub PAT válido para autenticar la petición."
        };
    }

    const endpointUrl = `https://api.github.com/repos/${owner}/${repo}/dispatches`;

    const payload = {
        event_type: "nueva_factura",
        client_payload: {
            pdf_url: pdfUrl,
            nombre_archivo: nombreArchivo
        }
    };

    const headers = {
        "Accept": "application/vnd.github.v3+json",
        "Authorization": `Bearer ${pat}`,
        "Content-Type": "application/json",
        "User-Agent": "SistemaFacturacionApp"
    };

    try {
        console.log(`[GitHub Service] 🚀 Enviando evento repository_dispatch para '${nombreArchivo}'...`);
        
        const response = await fetch(endpointUrl, {
            method: 'POST',
            headers: headers,
            body: JSON.stringify(payload)
        });

        if (response.status === 204) {
            const successMsg = `[GitHub Service] ✅ Evento 'nueva_factura' enviado con éxito para '${nombreArchivo}' (HTTP 204 No Content).`;
            console.log(successMsg);
            return {
                success: true,
                status: 204,
                message: successMsg
            };
        } else {
            const errorMsg = `[GitHub Service] ❌ Error al enviar dispatch a GitHub (${owner}/${repo}): HTTP ${response.status} ${response.statusText}`;
            console.error(errorMsg);
            return {
                success: false,
                status: response.status,
                message: errorMsg
            };
        }
    } catch (error) {
        const errorMsg = `[GitHub Service] ❌ Excepción de red/sistema al conectar con GitHub API: ${error.message || error}`;
        console.error(errorMsg, error);
        return {
            success: false,
            status: 0,
            message: errorMsg
        };
    }
}

/**
 * Envía una solicitud de dispatch a GitHub Actions para procesar TODO EL LOTE completo de facturas desde Supabase Storage en 1 solo ciclo.
 * 
 * @param {number} totalArchivos - Cantidad total de archivos subidos en el lote.
 * @returns {Promise<{success: boolean, status: number, message: string}>} Resultado de la operación.
 */
export async function enviarLoteAGitHub(totalArchivos) {
    const owner = DEFAULT_GITHUB_OWNER;
    const repo = DEFAULT_GITHUB_REPO;
    const pat = obtenerGitHubPAT();

    if (!pat) {
        return {
            success: false,
            status: 401,
            message: "No se proporcionó un GitHub PAT válido para autenticar la petición."
        };
    }

    const endpointUrl = `https://api.github.com/repos/${owner}/${repo}/dispatches`;

    const payload = {
        event_type: "procesar_lote",
        client_payload: {
            batch: true,
            total_archivos: totalArchivos
        }
    };

    const headers = {
        "Accept": "application/vnd.github.v3+json",
        "Authorization": `Bearer ${pat}`,
        "Content-Type": "application/json",
        "User-Agent": "SistemaFacturacionApp"
    };

    try {
        console.log(`[GitHub Service] 🚀 Enviando evento 'procesar_lote' a GitHub Actions para ${totalArchivos} facturas...`);
        
        const response = await fetch(endpointUrl, {
            method: 'POST',
            headers: headers,
            body: JSON.stringify(payload)
        });

        if (response.status === 204) {
            const successMsg = `[GitHub Service] ✅ Evento 'procesar_lote' enviado con éxito para ${totalArchivos} facturas (HTTP 204 No Content).`;
            console.log(successMsg);
            return {
                success: true,
                status: 204,
                message: successMsg
            };
        } else {
            const errorMsg = `[GitHub Service] ❌ Error al enviar dispatch de lote a GitHub (${owner}/${repo}): HTTP ${response.status} ${response.statusText}`;
            console.error(errorMsg);
            return {
                success: false,
                status: response.status,
                message: errorMsg
            };
        }
    } catch (error) {
        const errorMsg = `[GitHub Service] ❌ Excepción de red/sistema al conectar con GitHub API: ${error.message || error}`;
        console.error(errorMsg, error);
        return {
            success: false,
            status: 0,
            message: errorMsg
        };
    }
}

// Exportación por defecto para facilitar distintos tipos de importación
export default {
    enviarFacturaAGitHub,
    enviarLoteAGitHub
};
